### PR TITLE
Fix pulling in metrics from foreign varnishclusters in the built-in grafana dashboard.

### PR DESCRIPTION
**Pull request by tomashibm**
Inject service name as a filter to grafana queries in order to not pull metrics from foreign varnishclusters.

Resolves #20.

### CODE REVIEW

**Review by kscharm — APPROVED** [review]
LGTM assuming it works

**Comment by tomashibm** [comment]
Forgot about the edge case when two VarnishClusters can have the same name but live in different namespaces. Added an another filter parameter for namespace.
Also a few fixes for servicemonitors.
Please have an another look.

**Review by cin — COMMENTED** [review]
do we need the pod in the metric as well? i.e. is service/namespace good enough?

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go [hunk 1]
```diff
@@ -2342,20 +2341,26 @@ var grafanaDashboardTemplate =
       {
         "allValue": null,
         "current": {
+          "selected": false,
           "text": "All",
           "value": "$__all"
         },
-        "datasource": "{{.DatasourceName}}",
```
Comment (kscharm) [inline pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2348]: Here you change the `.DatasourceName` to `Prometheus-varnish`, but you don't remove the field from the `data` map. Did you mean to do this?

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go [hunk 2]
```diff
@@ -2342,20 +2341,26 @@ var grafanaDashboardTemplate =
...
+        "name": "varnish_pod",
         "options": [],
-        "query": "",
+        "query": {
+          "query": "label_values(varnish_up{ service=\"{{.ServiceName}}\", namespace=\"{{.Namespace}}\", namespace=\"{{.Namespace}}\" }, pod)",
+          "refId": "StandardVariableQuery"
+        },
         "refresh": 1,
-        "regex": "/^(.*):/",
+        "regex": "",
```
Comment (kscharm) [inline pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2363]: Also what is the reason for removing the regex?

**Review by kscharm — CHANGES_REQUESTED** [review]
A few things I have questions about

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go (hunk 1 above)
Comment (tomashibm) [reply pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2348]: No, I accidentally changed that. Reverted. Good catch.

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go (hunk 2 above)
Comment (tomashibm) [reply pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2363]: So before I needed the IP address of the pod, so I queried the `instance` label that has the IP and the port (e.g. `10.0.0.12:1232` -> `10.0.0.12`). That regexp extracted the IP. Now I need only the pod name which comes without anything else.

**Review by tomashibm — COMMENTED** [review]
Yeah, service + namespace should be good enough.

The pod variable can be also use to filter out metrics for individual varnish instances (ALL by default).

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go (hunk 1 above)
Comment (cin) [reply pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2348]: Indeed! good catch

**Comment by tomashibm** [comment]
@kscharm anything else you want me to address?

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go [hunk 3]
```diff
@@ -2284,7 +2283,7 @@ var grafanaDashboardTemplate =
       "steppedLine": false,
       "targets": [
         {
-          "expr": "sum(irate(varnish_main_fetch{}[1m])) by (type)",
+          "expr": "sum(irate(varnish_main_fetch{ service=\"{{.ServiceName}}\", namespace=\"{{.Namespace}}\", namespace=\"{{.Namespace}}\" }[1m])) by (type)",
```
Comment (kscharm) [inline pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2286]: Why is namespace specified twice here?

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go [hunk 4]
```diff
@@ -2342,20 +2341,26 @@ var grafanaDashboardTemplate =
       {
         "allValue": null,
         "current": {
+          "selected": false,
           "text": "All",
           "value": "$__all"
         },
-        "datasource": "{{.DatasourceName}}",
-        "definition": "label_values(varnish_up, instance)",
+        "datasource": "Prometheus-varnish",
```
Comment (kscharm) [inline pkg/varnishcluster/controller/varnishcluster_grafana_dashboard_template.go:2348]: Still seeing the incorrect datasource.

**Review by kscharm — CHANGES_REQUESTED** [review]
A few small issues.

File: pkg/varnishcluster/controller/varnishcluster_grafana_dashboard.go [hunk 5]
```diff
@@ -194,7 +194,9 @@ func (r *ReconcileVarnishCluster) createGrafanaDashboardConfigMap(instance *vcap
 func generateGrafanaDashboardData(instance *vcapi.VarnishCluster) (map[string]string, error) {
 	data := map[string]interface{}{
 		"DatasourceName": *instance.Spec.Monitoring.GrafanaDashboard.DatasourceName,
-		"Title": instance.Spec.Monitoring.GrafanaDashboard.Title,
+		"Title":          instance.Spec.Monitoring.GrafanaDashboard.Title,
+		"ServiceName":    instance.Name,
```
Comment (kscharm) [inline pkg/varnishcluster/controller/varnishcluster_grafana_dashboard.go:198]: Also why call it `ServiceName` when it's the name of the `VarnishCluster`?

Comment (tomashibm) [reply pkg/varnishcluster/controller/varnishcluster_grafana_dashboard.go:198]: It references the `service` label of the metric. And the Kubernetes service that exposes Varnish pods is called by the name of the VarnishCluster. It doesn't have an additional suffix.

**Review by kscharm — APPROVED** [review]
LGTM!

**Pull request merged**